kong/db/schema/topological_sort.lua

```diff
@@ -0,0 +1,106 @@
+local table_insert = table.insert
+
+-- Given an array of schemas, build a table were the keys are schemas and the values are
+-- the list of schemas with foreign keys pointing to them
+-- @tparam array schemas an array of schemas
+-- @treturn table a map with schemas in the keys and arrays of neighbors in the values
+-- @usage
+-- local res = build_neighbors_map({ services, routes, plugins, consumers })
+--
+-- assert.same({ [services] = { routes, plugins },
+--               [routes] = { plugins },
+--               [consumers] = { plugins }
+--             }, res)
+local function build_neighbors_map(schemas)
+  local schemas_by_name = {}
+  local s
+  for i = 1, #schemas do
+    s = schemas[i]
+    schemas_by_name[s.name] = s
+  end
+
+  local res = {}
+  local source, destination
+
+  for i = 1, #schemas do
+    source = schemas[i] -- routes
+    for _, field in source:each_field() do
+      if field.type == "foreign"  then
+        destination = schemas_by_name[field.reference] -- services
+        if destination then
+          res[destination] = res[destination] or {}
+          table_insert(res[destination], source)
+        end
+      end
+    end
+  end
+
+  return res
+end
+
+
+-- aux function for topological_sort
+local function visit(current, neighbors_map, visited, marked, sorted)
+  if visited[current] then
+    return true
+  end
+
+  if marked[current] then
+    return nil, "Cycle detected, can not sort topographically"
```

Comment:
`can not` -> `cannot`

Reply:
also, `topographically` -> `topologically`